app/api/dao/user_extension.py

```diff
@@ -97,20 +54,34 @@ def update_user_additional_info(data):
         timezone_value = data["timezone"]
         timezone = Timezone(timezone_value).name
         additional_info_data = {}
-        
-        existing_additional_info = UserExtensionModel.find_by_user_id(user_id)
-        if not existing_additional_info:
-            return messages.ADDITIONAL_INFORMATION_DOES_NOT_EXIST, HTTPStatus.NOT_FOUND
         try:
-            existing_additional_info.is_organization_rep = data["is_organization_rep"]
-            existing_additional_info.timezone = timezone
             additional_info_data["phone"] = data["phone"]
             additional_info_data["mobile"] = data["mobile"]
             additional_info_data["personal_website"] = data["personal_website"]
         except KeyError as e:
             return e, HTTPStatus.BAD_REQUEST
+        
+        existing_additional_info = UserExtensionModel.find_by_user_id(user_id)
+        if not existing_additional_info:
+            user_extension = UserExtensionModel(user_id, timezone)
+            try:
+                user_extension.is_organization_rep = data["is_organization_rep"]
+            except KeyError as e:
+                return e, HTTPStatus.BAD_REQUEST
 
```

Comment:
As part of refactoring, the code duplication can be abstracted to a separate method. 
eg. while updating you set organization, timezone, save to db for existing info otherwise if it's new then you do same stuff for user_extension. 
This could be abstracted, and it could be a method you call, which does these things

Reply:
@meenakshi-dhanani. I don't think it's a duplication. When the data does not exist, the a new instance is created with parameters pass to UserExtensionModel 
<img width="710" alt="Screen Shot 2020-08-04 at 8 59 10 am" src="https://user-images.githubusercontent.com/29667122/89235246-30482f00-d631-11ea-9c72-b53cc2c7d50a.png">


But when data exist which means user wants to update, the existing data attributes is replaced (not using the UserExtensionModel)

<img width="707" alt="Screen Shot 2020-08-04 at 9 03 17 am" src="https://user-images.githubusercontent.com/29667122/89235425-7bfad880-d631-11ea-9121-c9b6b4bae52c.png">


If you think this is still a duplication, can you please show me how to make the code better (to use one instance for replacing existing attributes in the data that can also be used when to represent a new instance)

Reply:
There is code duplication, if you see carefully. You can abstract is out, the following is an example of how you can abstract common logic out to a method. 

`def update_user_additional_info(data):
        """Updates a user_extension instance.
        Arguments:
            data: A list containing user's id, boolean value of whether or not
            the user is representing an organization, as well as their timezone
    
        Returns:
                A dictionary containing "message" which indicates whether or not 
                the user_exension was updated successfully and "code" for the HTTP response code.
        """

        try:
            user_id = int(AUTH_COOKIE["user_id"].value)
        except ValueError:
            return messages.USER_ID_IS_NOT_RETRIEVED, HTTPStatus.FORBIDDEN    

        
        existing_additional_info = UserExtensionModel.find_by_user_id(user_id)
        if not existing_additional_info:
            user_extension = UserExtensionModel(user_id, timezone)
            return update(user_extension, data, messages.ADDITIONAL_INFO_SUCCESSFULLY_CREATED, HTTPStatus.CREATED)    
        return update(existing_additional_info, data, messages.ADDITIONAL_INFO_SUCCESSFULLY_UPDATED, HTTPStatus.OK)     
      
    def update(user, data, success_message, status_code):
        additional_info_data = {}
        try:
            additional_info_data["phone"] = data["phone"]
            additional_info_data["mobile"] = data["mobile"]
            additional_info_data["personal_website"] = data["personal_website"]
        except KeyError as e:
            return e, HTTPStatus.BAD_REQUEST

        try: 
            user.is_organization_rep = data["is_organization_rep"]
        except KeyError as e:
            return e, HTTPStatus.BAD_REQUEST

        timezone_value = data["timezone"]
        timezone = Timezone(timezone_value).name            
        user.timezone = timezone
        user.additional_info = additional_info_data
        user.save_to_db()

        return success_message, status_code
`

In the update method, all common logic resides. So now same code is not repeated twice. And if you look at the update_user_additional_info method, see how readable it is when you say if a user_extension exists then update status and statuscode and other wise created and all. 

Does it make sense?

Reply:
Hi @meenakshi-dhanani . Thanks for your refactoring suggestion. Have you tried to run the application with your code change? I've tried but the code failed as the Timezone which is needed for initiating the UserExtension model is not defined before the initiation line.

<img width="984" alt="Screen Shot 2020-08-05 at 12 47 54 pm" src="https://user-images.githubusercontent.com/29667122/89366524-601f3180-d71a-11ea-9c4d-9475f4140bc4.png">

Reply:
no problem, feel free to remove the timezone logic from the method and put it up as it is. My idea was to give you something to figure out how to spot the duplication and extract into method. It doesn't have to be pasted as it is. I'd like if you don't paste it all. And try doing it once you get how to.

Reply:
Don't worry. I've fixed it by moving timezone before initiation code.

Reply:
`def update_user_additional_info(data):
        """Updates a user_extension instance.
        Arguments:
            data: A list containing user's id, boolean value of whether or not
            the user is representing an organization, as well as their timezone
    
        Returns:
                A dictionary containing "message" which indicates whether or not 
                the user_exension was updated successfully and "code" for the HTTP response code.
        """

        try:
            user_id = int(AUTH_COOKIE["user_id"].value)
        except ValueError:
            return messages.USER_ID_IS_NOT_RETRIEVED, HTTPStatus.FORBIDDEN    

        
        
        timezone_value = data["timezone"]
        timezone = Timezone(timezone_value).name            

        existing_additional_info = UserExtensionModel.find_by_user_id(user_id)
        if not existing_additional_info:
            user_extension = UserExtensionModel(user_id, timezone)
            return update(user_extension, data, messages.ADDITIONAL_INFO_SUCCESSFULLY_CREATED, HTTPStatus.CREATED)    
        response = update(existing_additional_info, data, messages.ADDITIONAL_INFO_SUCCESSFULLY_UPDATED, HTTPStatus.OK)
        existing_additional_info.timezone = timezone
        return response
      
    def update(user, data, success_message, status_code):
        additional_info_data = {}
        try:
            additional_info_data["phone"] = data["phone"]
            additional_info_data["mobile"] = data["mobile"]
            additional_info_data["personal_website"] = data["personal_website"]
        except KeyError as e:
            return e, HTTPStatus.BAD_REQUEST

        try: 
            user.is_organization_rep = data["is_organization_rep"]
        except KeyError as e:
            return e, HTTPStatus.BAD_REQUEST

        user.additional_info = additional_info_data
        user.save_to_db()

        return success_message, status_code
`

Reply:
Again, this is just a suggestion, I tried making the timezone common. But try yourself and see what works for you

Reply:
Thank again for the suggestion. But I think putting timezone after updating the database record as you suggested above will beat the purpose of updating the database on timexzone field. 
So, I'll keep it as my latest change for now.

Reply:
yes yes. I made the comment before seeing your change